edgelesssys/constellation
# operators: infrastructure autodiscovery

### Context

This change is part of the larger effort to support arbitrary node groups.
When new node groups are added or old node groups are deleted, the cluster needs to learn about those changes in order to correctly control the cluster-autoscaler and perform upgrades safely.
This PR adds two new controllers in order to react to infrastructure changes:

- NodeJoinWacher watches for nodes that join and leave the cluster. It's used to trigger a reconciliation of the other new controller
- ExternalScalingGroupReconciler finds scaling groups in cloud provider infrastructure and dynamically creates / deletes ScaleSet CRs


### Proposed change(s)

- operators: infrastructure autodiscovery
- docs: document scaling configuration when using multiple node groups





### How to test

- create a cluster on GCP
- manually add an additional worker node group in the terraform.tfvars file and `terraform apply` the changes
- wait for new nodes to join
- follow new docs and see if autoscaling works with all worker node groups

### Checklist
